modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/index/RocksDbSortedIndexStorage.java

```diff
@@ -167,24 +167,71 @@ private Cursor<ByteBuffer> createScanCursor(byte @Nullable [] lowerBound, byte @
 
         RocksIterator it = indexCf.newIterator(options);
 
-        if (lowerBound == null) {
-            it.seek(partitionStorage.partitionStartPrefix());
-        } else {
-            it.seek(lowerBound);
-        }
+        return new Cursor<>() {
+            @Nullable
+            private Boolean hasNext;
 
-        return new RocksIteratorAdapter<>(it) {
-            @Override
-            protected ByteBuffer decodeEntry(byte[] key, byte[] value) {
-                return ByteBuffer.wrap(key).order(ORDER);
-            }
+            @Nullable
+            private byte[] key;
 
             @Override
             public void close() {
-                super.close();
+                it.close();
 
                 RocksUtils.closeAll(options, upperBoundSlice);
             }
+
+            @Override
+            public boolean hasNext() {
+                advanceIfNeeded();
+
+                return hasNext;
+            }
+
+            @Override
+            public ByteBuffer next() {
+                advanceIfNeeded();
+
+                boolean hasNext = this.hasNext;
+
+                this.hasNext = null;
+
+                if (!hasNext) {
+                    throw new NoSuchElementException();
+                }
+
+                return ByteBuffer.wrap(key).order(ORDER);
+            }
+
+            private void advanceIfNeeded() throws StorageException {
+                if (hasNext == null) {
+                    try {
+                        it.refresh();
+                    } catch (RocksDBException e) {
+                        throw new StorageException("Error refreshing an iterator", e);
+                    }
+
+                    if (key == null) {
+                        it.seek(lowerBound == null ? partitionStorage.partitionStartPrefix() : lowerBound);
+                    } else {
+                        it.seekForPrev(key);
+
+                        it.next();
```

Comment:
Technically, iterator may not be valid in this case and calling `next` is unsafe. You should add a test where index has a single value and it's deleted while iterator is "standing" on it.
It'll most likely fail.

Reply:
Tried to test in `org.apache.ignite.internal.storage.index.AbstractSortedIndexStorageTest#testScanContractRemoveCachedRow`